cmd/setup-etcd-environment: refactor and add tests

hexfusion opened the pull request:
During the work before feature freeze, we introduced a regression in DR. The codebase had become a bit convoluted and hard to test. This PR introduces new unit tests that cover DR to prevent this from happening again. As we continue efforts with `cluster-etcd-operator` these e2e tests will be extended include that functionality.

The regression was reverted in 4.3 FTR [1]

[1] https://github.com/openshift/machine-config-operator/pull/1304

hexfusion commented:
/test e2e-aws-disruptive

hexfusion commented:
/skip
/retest

hexfusion commented:
/skip

wking commented on cmd/setup-etcd-environment/run_test.go:
```diff
@@ -16,164 +138,231 @@ import (
 
 	if bootEnv["DISCOVERY_SRV"] != "" {
-		t.Errorf("bootEnv[\"DISCOVERY_SRV\"] = \"%s\"; want \"\" ", bootEnv["DISCOVERY_SRV"])
+		t.Errorf("bootEnv[\"DISCOVERY_SRV\"] = (%s); want (%s)", bootEnv["DISCOVERY_SRV"], "")
```
nit, `%q`?  But paren-wrapping is probably safe enough too :p

hexfusion replied:
good point updated!

hexfusion commented:
/test e2e-gcp-upgrade

retroflexer commented on cmd/setup-etcd-environment/run.go:
```diff
@@ -36,22 +39,78 @@ var (
 	rootCmd.PersistentFlags().BoolVar(&runOpts.bootstrapSRV, "bootstrap-srv", true, "use SRV discovery for bootstraping etcd cluster.")
 }
 
+func NewSetupEnv(runOpts *opts, etcdName, etcdDataDir string, ips []string) (*SetupEnv, error) {
```
minor nit: does this need to be a public function? (this is a self-contained executable, so it perhaps doesn't matter!

retroflexer commented:
/lgtm

hexfusion commented:
/retest

hexfusion commented:
/test e2e-gcp-upgrade

hexfusion commented:
last upgrade fail :/
```
Nov 26 14:40:45.334 W ns/openshift-machine-config-operator pod/etcd-quorum-guard-8987d47f-kx2jn 0/6 nodes are available: 2 node(s) didn't match node selector, 2 node(s) didn't match pod affinity/anti-affinity, 2 node(s) didn't satisfy existing pods anti-affinity rules, 2 node(s) were unschedulable.
```

kikisdeliveryservice commented:
/skip

kikisdeliveryservice commented on cmd/setup-etcd-environment/run.go:
```diff
@@ -129,53 +164,82 @@ func runRunCmd(cmd *cobra.Command, args []string) error {
-		"DNS_NAME":          dns,
-		"WILDCARD_DNS_NAME": fmt.Sprintf("*.%s", runOpts.discoverySRV),
-	}, out, false)
+func (s *SetupEnv) getExportEnv() (map[string]string, error) {
```
dumb q but why Export?

hexfusion replied:
We export certain variables so that the ENV is available not only to the container but also to the binary involved by the container as well.

kikisdeliveryservice replied:
i think the func name really wasn't clicking with my brain. makes sense!

kikisdeliveryservice commented on cmd/setup-etcd-environment/run.go:
```diff
@@ -129,53 +164,82 @@ func runRunCmd(cmd *cobra.Command, args []string) error {
+	env["NAME"] = s.etcdName
+
+	memberList := s.getEtcdMemberList()
+	env["INITIAL_CLUSTER"] = strings.Join(memberList, ",")
```
this ends up being not an array but a string: "blah, blah, blah" ?

wondering why this is a string and not an array (prob a good reason, but would like to understand)

retroflexer replied:
This is typical of the INITIAL_CLUSTER environment variable to have a comma separated string of member list. For instance, we build the INITIAL_CLUSTER when we restore a backup we set the INITIAL_CLUSTER as:
echo $INITIAL_CLUSTER
etcd-member-ip-10-0-131-52.ec2.internal=https://etcd-1.alpatel-test-3.devcluster.openshift.com:2380,etcd-member-ip-10-0-157-131.ec2.internal=https://etcd-1.alpatel-test-3.devcluster.openshift.com:2380,etcd-member-ip-10-0-164-217.ec2.internal=https://etcd-2.alpatel-test-3.devcluster.openshift.com:2380

See https://docs.openshift.com/container-platform/4.1/backup_and_restore/disaster_recovery/scenario-2-restoring-cluster-state.html (section 1.c)

hexfusion replied:
Sure so ETCD_INITIAL_CLUSTER [1] is a comma-delimited string [2] that accepts peers in the following format `<name>=<peerURL>`.

[1] https://github.com/etcd-io/etcd/blob/a621d807f061e1dd635033a8d6bc261461429e27/Documentation/op-guide/configuration.md#--initial-cluster

[2] https://github.com/etcd-io/etcd/blob/a65e4d3357c260504d47f08a433b355e36dbcb51/embed/config.go#L198

kikisdeliveryservice replied:
ahhhh TIL.. thank you both for explaining and linking! :)

kikisdeliveryservice commented on cmd/setup-etcd-environment/run.go:
```diff
@@ -129,53 +164,82 @@ func runRunCmd(cmd *cobra.Command, args []string) error {
+	for _, member := range hostEtcdEndpoint.Subsets[0].Addresses {
+		// TODO validate IP?
+		if member.Hostname == etcdBootstrapName {
+			endpoints = append(endpoints, fmt.Sprintf("https://%s:2379", member.IP))
```
will this fail if member.IP is empty/nil?  maybe check first

hexfusion replied:
yeah I think it is a loose assumption that has value, good catch.

kikisdeliveryservice commented:
/skip

hexfusion commented:
/retest

hexfusion commented:
/retest

hexfusion commented:
/retest

Going to dive into the failure.

hexfusion commented:
> /retest
> 
> Going to dive into the failure.

etcd is coming up fine and as these changes are isolated to the init container discovery which is a dependency for etcd I am not finding any reason that these are related.

hexfusion commented:
/restest

hexfusion commented:
> [systemd]
>Failed Units: 2
>  machine-config-daemon-firstboot.service
>  machine-config-daemon-host.service

In libvirt if I ssh into the master I see the above @kikisdeliveryservice

hexfusion commented:
> journalctl | grep machine-config-daemon-firstboot.service
Dec 03 00:57:35 sbatsche-xvqbv-master-0.sbatsche.hexfusion.local ignition[1124]: INFO     : files: op(2b): [started]  processing unit "machine-config-daemon-firstboot.service"
Dec 03 00:57:35 sbatsche-xvqbv-master-0.sbatsche.hexfusion.local ignition[1124]: INFO     : files: op(2b): op(2c): [started]  writing unit "machine-config-daemon-firstboot.service" at "etc/systemd/system/machine-config-daemon-firstboot.service"
Dec 03 00:57:35 sbatsche-xvqbv-master-0.sbatsche.hexfusion.local ignition[1124]: INFO     : files: op(2b): op(2c): [finished] writing unit "machine-config-daemon-firstboot.service" at "etc/systemd/system/machine-config-daemon-firstboot.service"
Dec 03 00:57:35 sbatsche-xvqbv-master-0.sbatsche.hexfusion.local ignition[1124]: INFO     : files: op(2b): [finished] processing unit "machine-config-daemon-firstboot.service"
Dec 03 00:57:35 sbatsche-xvqbv-master-0.sbatsche.hexfusion.local ignition[1124]: INFO     : files: op(2d): [started]  enabling unit "machine-config-daemon-firstboot.service"
Dec 03 00:57:35 sbatsche-xvqbv-master-0.sbatsche.hexfusion.local ignition[1124]: INFO     : files: op(2d): [finished] enabling unit "machine-config-daemon-firstboot.service"
Dec 03 00:57:38 sbatsche-xvqbv-master-0.sbatsche.hexfusion.local restorecon[1597]: Relabeled /etc/systemd/system/machine-config-daemon-firstboot.service from system_u:object_r:unlabeled_t:s0 to system_u:object_r:systemd_unit_file_t:s0
Dec 03 00:57:39 sbatsche-xvqbv-master-0.sbatsche.hexfusion.local rpm-ostree[1795]: client(id:cli dbus:1.4 unit:machine-config-daemon-firstboot.service uid:0) added; new total=2
Dec 03 00:57:39 sbatsche-xvqbv-master-0.sbatsche.hexfusion.local rpm-ostree[1795]: client(id:cli dbus:1.4 unit:machine-config-daemon-firstboot.service uid:0) vanished; remaining=1
Dec 03 00:57:39 sbatsche-xvqbv-master-0 machine-config-daemon[1737]: client(id:cli dbus:1.4 unit:machine-config-daemon-firstboot.service uid:0) added; new total=2
Dec 03 00:57:39 sbatsche-xvqbv-master-0 machine-config-daemon[1737]: client(id:cli dbus:1.4 unit:machine-config-daemon-firstboot.service uid:0) vanished; remaining=1
Dec 03 00:58:00 sbatsche-xvqbv-master-0 systemd[1]: machine-config-daemon-firstboot.service: Main process exited, code=exited, status=1/FAILURE
Dec 03 00:58:00 sbatsche-xvqbv-master-0 systemd[1]: machine-config-daemon-firstboot.service: Failed with result 'exit-code'.
Dec 03 00:58:00 sbatsche-xvqbv-master-0 systemd[1]: machine-config-daemon-firstboot.service: Consumed 111ms CPU time

hexfusion commented:
```
$ journalctl | grep machine-config-daemon-host.service
Dec 03 00:57:34 sbatsche-xvqbv-master-0.sbatsche.hexfusion.local ignition[1124]: INFO     : files: createFilesystemsFiles: createFiles: op(a): [started]  writing file "/etc/systemd/system/machine-config-daemon-host.service.d/10-default-env.conf"
Dec 03 00:57:34 sbatsche-xvqbv-master-0.sbatsche.hexfusion.local ignition[1124]: INFO     : files: createFilesystemsFiles: createFiles: op(a): [finished] writing file "/etc/systemd/system/machine-config-daemon-host.service.d/10-default-env.conf"
Dec 03 00:57:35 sbatsche-xvqbv-master-0.sbatsche.hexfusion.local ignition[1124]: INFO     : files: op(2e): [started]  processing unit "machine-config-daemon-host.service"
Dec 03 00:57:35 sbatsche-xvqbv-master-0.sbatsche.hexfusion.local ignition[1124]: INFO     : files: op(2e): op(2f): [started]  writing unit "machine-config-daemon-host.service" at "etc/systemd/system/machine-config-daemon-host.service"
Dec 03 00:57:35 sbatsche-xvqbv-master-0.sbatsche.hexfusion.local ignition[1124]: INFO     : files: op(2e): op(2f): [finished] writing unit "machine-config-daemon-host.service" at "etc/systemd/system/machine-config-daemon-host.service"
Dec 03 00:57:35 sbatsche-xvqbv-master-0.sbatsche.hexfusion.local ignition[1124]: INFO     : files: op(2e): [finished] processing unit "machine-config-daemon-host.service"
Dec 03 00:57:35 sbatsche-xvqbv-master-0.sbatsche.hexfusion.local ignition[1124]: INFO     : files: op(30): [started]  enabling unit "machine-config-daemon-host.service"
Dec 03 00:57:35 sbatsche-xvqbv-master-0.sbatsche.hexfusion.local ignition[1124]: INFO     : files: op(30): [finished] enabling unit "machine-config-daemon-host.service"
Dec 03 00:57:38 sbatsche-xvqbv-master-0.sbatsche.hexfusion.local restorecon[1597]: Relabeled /etc/systemd/system/machine-config-daemon-host.service.d from system_u:object_r:unlabeled_t:s0 to system_u:object_r:systemd_unit_file_t:s0
Dec 03 00:57:38 sbatsche-xvqbv-master-0.sbatsche.hexfusion.local restorecon[1597]: Relabeled /etc/systemd/system/machine-config-daemon-host.service.d/10-default-env.conf from system_u:object_r:unlabeled_t:s0 to system_u:object_r:systemd_unit_file_t:s0
Dec 03 00:57:38 sbatsche-xvqbv-master-0.sbatsche.hexfusion.local restorecon[1597]: Relabeled /etc/systemd/system/machine-config-daemon-host.service from system_u:object_r:unlabeled_t:s0 to system_u:object_r:systemd_unit_file_t:s0
Dec 03 00:57:39 sbatsche-xvqbv-master-0.sbatsche.hexfusion.local rpm-ostree[1795]: client(id:cli dbus:1.3 unit:machine-config-daemon-host.service uid:0) added; new total=1
Dec 03 00:57:39 sbatsche-xvqbv-master-0.sbatsche.hexfusion.local rpm-ostree[1795]: client(id:cli dbus:1.3 unit:machine-config-daemon-host.service uid:0) vanished; remaining=0
Dec 03 00:57:39 sbatsche-xvqbv-master-0 machine-config-daemon[1737]: client(id:cli dbus:1.3 unit:machine-config-daemon-host.service uid:0) added; new total=1
Dec 03 00:57:39 sbatsche-xvqbv-master-0 machine-config-daemon[1737]: client(id:cli dbus:1.3 unit:machine-config-daemon-host.service uid:0) vanished; remaining=0
Dec 03 00:58:00 sbatsche-xvqbv-master-0 systemd[1]: machine-config-daemon-host.service: Main process exited, code=exited, status=1/FAILURE
Dec 03 00:58:00 sbatsche-xvqbv-master-0 systemd[1]: machine-config-daemon-host.service: Failed with result 'exit-code'.
Dec 03 00:58:00 sbatsche-xvqbv-master-0 systemd[1]: machine-config-daemon-host.service: Consumed 21.133s CPU time
Dec 03 00:58:00 sbatsche-xvqbv-master-0 machine-config-daemon[1737]: error: failed to run pivot: failed to start machine-config-daemon-host.service: exit status 1
Dec 03 01:06:48 sbatsche-xvqbv-master-0 hyperkube[2115]: I1203 01:06:48.330097    2115 factory.go:170] Factory "raw" can handle container "/system.slice/machine-config-daemon-host.service", but ignoring.
Dec 03 01:06:48 sbatsche-xvqbv-master-0 rpm-ostree[22302]: client(id:cli dbus:1.493 unit:machine-config-daemon-host.service uid:0) added; new total=1
Dec 03 01:06:48 sbatsche-xvqbv-master-0 rpm-ostree[22302]: client(id:cli dbus:1.493 unit:machine-config-daemon-host.service uid:0) vanished; remaining=0
Dec 03 01:07:04 sbatsche-xvqbv-master-0 systemd[1]: machine-config-daemon-host.service: Main process exited, code=exited, status=1/FAILURE
Dec 03 01:07:04 sbatsche-xvqbv-master-0 systemd[1]: machine-config-daemon-host.service: Failed with result 'exit-code'.
Dec 03 01:07:04 sbatsche-xvqbv-master-0 systemd[1]: machine-config-daemon-host.service: Consumed 1.037s CPU time
Dec 03 01:07:48 sbatsche-xvqbv-master-0 hyperkube[2115]: I1203 01:07:48.410809    2115 factory.go:170] Factory "raw" can handle container "/system.slice/machine-config-daemon-host.service", but ignoring.
Dec 03 01:07:48 sbatsche-xvqbv-master-0 rpm-ostree[22302]: client(id:cli dbus:1.567 unit:machine-config-daemon-host.service uid:0) added; new total=1
Dec 03 01:07:48 sbatsche-xvqbv-master-0 rpm-ostree[22302]: client(id:cli dbus:1.567 unit:machine-config-daemon-host.service uid:0) vanished; remaining=0
Dec 03 01:07:52 sbatsche-xvqbv-master-0 systemd[1]: machine-config-daemon-host.service: Main process exited, code=exited, status=1/FAILURE
Dec 03 01:07:52 sbatsche-xvqbv-master-0 systemd[1]: machine-config-daemon-host.service: Failed with result 'exit-code'.
Dec 03 01:07:52 sbatsche-xvqbv-master-0 systemd[1]: machine-config-daemon-host.service: Consumed 993ms CPU time
```

hexfusion commented:
/retest

alaypatel07 commented on cmd/setup-etcd-environment/run.go:
```diff
@@ -27,6 +27,10 @@ import (
 	EtcdScalingAnnotationKey = "etcd.operator.openshift.io/scale"
 	etcdInitialExisting      = "existing"
+	retryDuration            = 10 * time.Second
+	testIPAddress            = "127.0.0.256"
```
can we have test-related constants in the test file to unclutter this?

hexfusion replied:
sure

alaypatel07 commented on cmd/setup-etcd-environment/run.go:
```diff
@@ -27,6 +27,10 @@ import (
 	etcdInitialExisting      = "existing"
+	retryDuration            = 10 * time.Second
+	testIPAddress            = "127.0.0.256"
+	testHost                 = "etcd-test.testcluster.openshift.com"
```
move to test file again for locality?

alaypatel07 commented on cmd/setup-etcd-environment/run.go:
```diff
@@ -36,22 +40,78 @@ var (
+
+	if inCluster() {
+		wait.PollInfinite(retryDuration, func() (bool, error) {
+			if _, err := os.Stat("/var/run/secrets/kubernetes.io/serviceaccount/token"); os.IsNotExist(err) {
```
set `/var/run/secrets/kubernetes.io/serviceaccount/token` as const?

alaypatel07 commented on cmd/setup-etcd-environment/run.go:
```diff
@@ -129,53 +164,82 @@ func runRunCmd(cmd *cobra.Command, args []string) error {
+	env["INITIAL_CLUSTER"] = strings.Join(memberList, ",")
+	env["INITIAL_CLUSTER_STATE"] = etcdInitialExisting
+
+	endpoints, err := s.getEtcdEndpoints()
```
We probably need to have something here that verifies the member list we got [here](https://github.com/openshift/machine-config-operator/pull/1287/files#diff-f26a1d5dc0cdd6c1f8863a9a3d7a3630R194) with the endpoint list and prove a one on one correspondence. 

Can we have a TODO here and corresponding issue/trello card to track it for a follow-up PR?

hexfusion replied:
>We probably need to have something here that verifies the member list we got here with the endpoint list and prove a one on one correspondence.

Why?

If this list[1] is good enough for apiserver it better be good enough for us. we have a client balancer to handle failover if a member is not valid/down etc. If the list were not the same would we fail to bootstrap? I think this type of validation belongs in the operator. 

[1] https://github.com/openshift/machine-config-operator/pull/1287/files#diff-f26a1d5dc0cdd6c1f8863a9a3d7a3630R211

hexfusion replied:
I see your point that it is important if we have a bug but we need to guarantee the list is valid before we ship it IMO and that is on the operator.

hexfusion replied:
We have changes coming to the operator to allow this validation we can talk about that later.

alaypatel07 replied:
I agree the validation should be on the cluster-etcd-operator. Feel free to mark this resolved

alaypatel07 commented on cmd/setup-etcd-environment/run.go:
```diff
@@ -193,70 +257,51 @@ func setExportEnv(etcdName, dns string) (map[string]string, error) {
 			rkey := fmt.Sprintf("ETCD_%s", val)
-			if recoveryEnv[rkey] != "" {
-				bootstrapEnv[val] = recoveryEnv[rkey]
+			if existingEnv[rkey] != "" {
```
Curious at what is happening here with respect to this env variables?

In all test cases we test for `""` value of these variables, but we still set a non-nil values, this is getting me confused.

Would it make sense to have a comment to describe what value is expected in normal bootstrap and disaster recovery scenarios for better readability and debugging?

hexfusion replied:
I think I can document this better, sure.

hexfusion replied:
@retroflexer do you want to comment on the test cases?

retroflexer replied:
@alaypatel07 here is my understanding. With CEO added, now we have three different scenarios where we are setting up the export environment:
1. Normal bootstrap scenario: we should always have `DISCOVERY_SRV` 
2. Disaster Recovery case: we want to retain `INITIAL_CLUSTER` and `INITIAL_CLUSTER_STATE`. Also we need to make sure we do not include `DISCOVERY_SRV`
3. CEO is scaling: We make sure we retain `INITIAL_CLUSTER` and `INITIAL_CLUSTER_STATE`; also build the `ENDPOINTS` by querying.

hexfusion replied:
I agree with @retroflexer, I think the logic is strait forward enough if we need to handle any additional cases down the road we can revisit. Marking resolved

hexfusion commented on cmd/setup-etcd-environment/run.go:
```diff
@@ -67,56 +127,31 @@ func runRunCmd(cmd *cobra.Command, args []string) error {
+	// found we will use it to populate `ETCD_IPV4_ADDRESS`.
+	overrideIP := os.Getenv("ETCD_IPV4_ADDRESS")
+
+	ips, err := ipAddrs(overrideIP)
```
The reason we go through that process is that the value is arbitrary. Although in our use case we pass the value from Kube it does not stop someone from using an IP that is not part of the system in ENV. I think the sanity check is worth the effort in this case.

alaypatel07 replied:
hrm, that makes sense, thanks

hexfusion commented on cmd/setup-etcd-environment/run.go:
```diff
@@ -67,56 +127,31 @@ func runRunCmd(cmd *cobra.Command, args []string) error {
-	ips, err := ipAddrs()
+	// overrideIP allows for a specific IP to be used as long as it is found on the interface. If
+	// found we will use it to populate `ETCD_IPV4_ADDRESS`.
+	overrideIP := os.Getenv("ETCD_IPV4_ADDRESS")
```
I think `preferredIP` is probably more strait forward here.

hexfusion commented on cmd/setup-etcd-environment/run.go:
```diff
@@ -36,22 +40,78 @@ var (
+		})
+		clientConfig, err := rest.InClusterConfig()
+		if err != nil {
+			panic(err.Error())
```
IMO we should explode in fire if we are `inCluster`[1] and fail to create a client. I would like to a BZ for this but I am open to argument.

[1] https://github.com/openshift/machine-config-operator/blob/9efe4acd407e28ace65043d40179f6040f140e5d/cmd/setup-etcd-environment/run.go#L400

alaypatel07 replied:
I am not sure, I saw the other places are returning an error (Line 107) and only this one is panicking, was curious if there was a reason...

hexfusion replied:
yeah true I am just going to send error for now..

hexfusion commented on cmd/setup-etcd-environment/run.go:
```diff
@@ -280,20 +332,21 @@ func ipAddrs() ([]string, error) {
 		if !ip.IsGlobalUnicast() {
 			continue // we only want global unicast address
 		}
+		if overrideIP != "" {
```
update naming

hexfusion replied:
this does need some more thought I am going to update this a bit

hexfusion commented:
/hold 

waiting for https://github.com/openshift/machine-config-operator/pull/1211

hexfusion commented:
/retest

kikisdeliveryservice commented:
@hexfusion our tests are currently busted. trying to get to the bottom of it...

alaypatel07 commented on cmd/setup-etcd-environment/run.go:
```diff
@@ -283,17 +340,21 @@ func ipAddrs() ([]string, error) {
+				glog.Infof("ipAddrs: preferredIP %s found", preferredIP)
+				return []string{preferredIP}, nil
+			}
+			continue
```
drop?

hexfusion commented:
/retest

hexfusion commented:
> level=error msg="Error: Error applying IAM policy to project \"openshift-gce-devel-ci\": Too many conflicts.  Latest error: Error setting IAM policy for project \"openshift-gce-devel-ci\": googleapi: Error 409: There were concurrent policy changes. Please retry the whole read-modify-write with exponential backoff., aborted"

/test e2e-gcp-upgrade

hexfusion commented:
> level=info msg="Cluster operator console Progressing is True with SyncLoopRefreshProgressingInProgress: SyncLoopRefreshProgressing: Working toward version 0.0.1-2019-12-05-000529"
level=info msg="Cluster operator console Available is False with DeploymentAvailableFailedUpdate: DeploymentAvailable: 2 replicas ready at version 0.0.1-2019-12-05-000529"
level=info msg="Cluster operator insights Disabled is False with : "
level=info msg="Cluster operator support Disabled is True with NotAuthorized: Reporting was not allowed: your Red Hat account is not enabled for remote support or your token has expired"

Little too slow I guess..

/test e2e-gcp-upgrade

hexfusion commented:
> level=info msg="API v1.16.2 up"

etcd is bootstrapping fine..

/test e2e-gcp-upgrade

hexfusion commented:
> Dec 05 13:48:12.217 I ns/openshift-kube-apiserver-operator deployment/kube-apiserver-operator (combined from similar events): Status for clusteroperator/kube-apiserver changed: Degraded message changed from "NodeControllerDegraded: All master node(s) are ready\nEncryptionKeyControllerDegraded: rpc error: code = Unavailable desc = etcdserver: leader changed" to "StaticPodsDegraded: etcdserver: request timed out\nNodeControllerDegraded: All master node(s) are ready\nEncryptionKeyControllerDegraded: rpc error: code = Unavailable desc = etcdserver: leader changed" (9 times)

This test is failing because of the overhead of etcd encryption we need to boot resources.

/test e2e-gcp-upgrade

hexfusion commented:
Marketplace flake 

/test e2e-aws

retroflexer commented:
/lgtm

hexfusion commented:
level=info msg="Destroying the bootstrap resources..."
level=info msg="Waiting up to 30m0s for the cluster at https://api.ci-op-8kncs1nd-1354f.origin-ci-int-gce.dev.openshift.com:6443 to initialize..."
level=error msg="Cluster operator authentication Degraded is True with RouteHealthDegradedFailedGet: RouteHealthDegraded: failed to GET route: dial tcp 34.74.199.73:443: connect: connection refused"
level=info msg="Cluster operator authentication Progressing is Unknown with NoData: "
level=info msg="Cluster operator authentication Available is Unknown with NoData: "
level=info msg="Cluster operator insights Disabled is False with : "
level=info msg="Cluster operator support Disabled is True with NotAuthorized: Reporting was not allowed: your Red Hat account is not enabled for remote support or your token has expired"
level=fatal msg="failed to initialize the cluster: Cluster operator authentication is still updati

https://bugzilla.redhat.com/show_bug.cgi?id=1765280

flake

/test e2e-gcp-op

alaypatel07 commented on cmd/setup-etcd-environment/run.go:
```diff
@@ -129,53 +152,86 @@ func runRunCmd(cmd *cobra.Command, args []string) error {
-		return err
+	if setupEnv.etcdDNS != "" {
+		unexportedEnv["DNS_NAME"] = setupEnv.etcdDNS
 	}
```
Can we please have a log entry here so we see the populated env in CI?

https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/1287/pull-ci-openshift-machine-config-operator-master-e2e-gcp-op/579/artifacts/e2e-gcp-op/pods/openshift-etcd_etcd-member-ci-op-qfg5r-m-1.c.openshift-gce-devel-ci.internal_discovery.log

alaypatel07 replied:
Bot exported and unexported env variables please

hexfusion replied:
I think this needs to be done in the next PR extending tests. I will note TODO.

alaypatel07 commented on cmd/setup-etcd-environment/run.go:
```diff
@@ -193,70 +249,79 @@ func setExportEnv(etcdName, dns string) (map[string]string, error) {
+}
+
+func (s *SetupEnv) setClient() error {
+	if inCluster() && !s.opts.bootstrapSRV {
```
Not sure about this but can we

s/inCluster() && !s.opts.bootstrapSRV/inCluster() 

and push the check `!s.opts.bootstrapSRV/inCluster()` in function `inCluster`?

hexfusion replied:
we can consider this as TODO

hexfusion commented:
/skip

hexfusion commented:
/hold cancel

ipv6 work will continue in a followup PR

hexfusion commented:
@kikisdeliveryservice PTAL

hexfusion commented:
/hold

to coordinate with other PRs.

hexfusion commented:
/hold cancel

hexfusion commented:
/test e2e-gcp-upgrade

smarterclayton commented:
Looks like the new failure mode is node's don't rejoin

https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/logs/release-openshift-origin-installer-e2e-aws-disruptive-4.4/12
